cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/UsageHandler.java

```diff
@@ -110,7 +119,136 @@ public void getDatasetAppUsage(HttpRequest request, HttpResponder responder,
                                  @PathParam("namespace-id") String namespaceId,
                                  @PathParam("dataset-id") String datasetId) {
     final DatasetId id = new DatasetId(namespaceId, datasetId);
-    Set<? extends EntityId> ids = registry.getPrograms(id);
-    responder.sendJson(HttpResponseStatus.OK, ids);
+    Set<ProgramId> ids = registry.getPrograms(id);
+    responder.sendJson(HttpResponseStatus.OK,
+                       UsageHandlerBackwardCompatibility.IdCompatibleEntityIds.IdProgram.transform(ids));
+  }
+
+  /**
+   * Class to wrap stuff needed to make {@link UsageHandler} backward compatible. See CDAP-7316
+   * Also, see {@link IdCompatibleEntityIds} documentation for more details.
+   */
+  public static final class UsageHandlerBackwardCompatibility {
+
+    /**
+     * Just a simple POJO to support backward compatibility for {@link UsageHandler} See: CDAP-7316
+     * This class allows {@link UsageHandler} to return results in deprecated {@link Id} serialized form.
+     */
+    public abstract static class IdCompatibleEntityIds implements EntityIdCompatible {
```

Comment:
is this class needed? Couldn't we just have `BackwardCompatibility.IdDatasetInstance` instead of `BackwardCompatibility.IdCompatibleEntityIds.IdDatasetInstance`?